lib/refinement/changeset.rb

```diff
@@ -0,0 +1,211 @@
+        break unless dirs.add?(path)
+        add[path.dirname]
+      }
+      modifications.each do |mod|
+        add[mod.path.dirname]
+        add[mod.prior_path.dirname] if mod.prior_path
+      end
+      modifications +
+        dirs.map { |d| FileModification.new(path: Pathname("#{d}/").freeze, type: FileModification::DIRECTORY_CHANGE_TYPE) }
+    end
+
+    # @return [FileModification,Nil] the changeset for the given absolute path,
+    #   or `nil` if the given path is un-modified
+    # @param absolute_path [Pathname]
+    def include_path?(absolute_path:)
+      modified_absolute_paths[absolute_path]
+    end
+
+    # @return [Array<String>] An array of patterns converted from a
+    #         {Dir.glob} pattern to patterns that {File.fnmatch} can handle.
+    #         This is used by the {#relative_glob} method to emulate
+    #         {Dir.glob}.
+    #
+    #   The expansion provides support for:
+    #
+    #   - Literals
+    #
+    #       dir_glob_equivalent_patterns('{file1,file2}.{h,m}')
+    #       => ["file1.h", "file1.m", "file2.h", "file2.m"]
+    #
+    #   - Matching the direct children of a directory with `**`
+    #
+    #       dir_glob_equivalent_patterns('Classes/**/file.m')
+    #       => ["Classes/**/file.m", "Classes/file.m"]
+    #
+    # @param [String] pattern   A {Dir#glob} like pattern.
+    #
+    def dir_glob_equivalent_patterns(pattern)
+      pattern = pattern.gsub('/**/', '{/**/,/}')
+      values_by_set = {}
+      pattern.scan(/\{[^}]*\}/) do |set|
+        values = set.gsub(/[{}]/, '').split(',')
+        values_by_set[set] = values
+      end
+
+      if values_by_set.empty?
+        [pattern]
+      else
+        patterns = [pattern]
+        values_by_set.each do |set, values|
+          patterns = patterns.flat_map do |old_pattern|
+            values.map do |value|
+              old_pattern.gsub(set, value)
+            end
+          end
+        end
+        patterns
+      end
+    end
+    private :dir_glob_equivalent_patterns
+
+    # @return [FileModification,Nil] the changeset for the given absolute glob,
+    #   or `nil` if no files matching the glob were modified
+    # @param absolute_glob [String] a glob pattern for absolute paths, suitable for an invocation of `Dir.glob`
+    def include_glob?(absolute_glob:)
+      absolute_globs = dir_glob_equivalent_patterns(absolute_glob)
+      _path, modification = modified_absolute_paths.find do |absolute_path, _modification|
+        absolute_globs.any? do |glob|
+          File.fnmatch?(glob, absolute_path, File::FNM_CASEFOLD | File::FNM_PATHNAME)
+        end
+      end
+      modification
+    end
+
+    # @return [FileModification|Nil, String|Nil] a modification and yaml diff for the keypath at the given absolute path,
+    #  or `nil` if the value at the given keypath is un-modified
+    # @param absolute_path [Pathname]
+    # @param keypath [Array]
+    def include_yaml_keypath?(absolute_path:, keypath:)
```

Comment:
same here, again unless I have the Ruby-isms wrong with respect to these method names.